File: dynamic_programming/recursive_rod_cutting.py
import sys


def cut_rod(prices, length):
    if length == 0:
        return 0

    result = -sys.maxsize
    for i in range(1, length + 1):
        result = max(result, prices[i] + cut_rod(prices, length - i))

    return result


if __name__ == '__main__':
    prices = [0, 1, 5, 8, 9, 17, 17, 20, 24, 30]

    for i in range(len(prices)):
        result = cut_rod(prices, i)

        print(f"length = {i} / result = {result}")

File: dynamic_programming/recursive_rod_cutting.go
//go:build ignore

package main

import (
	"fmt"
	"math"
)

func cutRod(n int, prices []int) int {
	if n == 0 {
		return 0
	}

	result := -math.MaxInt
	for i := 1; i <= n; i++ {
		result = max(result, prices[i]+cutRod(n-i, prices))
	}

	return result
}

func main() {
	prices := []int{0, 1, 5, 8, 9, 10, 17, 17, 20, 24, 30}

	for i := 0; i < len(prices); i++ {
		result := cutRod(i, prices)

		fmt.Printf("length = %d / result = %d\n", i, result)
	}
}
